src/realm/sync/client.cpp

```diff
@@ -1103,24 +1099,53 @@ SyncClientHookAction SessionImpl::call_debug_hook(SyncClientHookEvent event, con
     return call_debug_hook(data);
 }
 
-bool SessionImpl::is_steady_state_download_message(DownloadBatchState batch_state, int64_t query_version)
+bool SessionImpl::is_steady_state_download_message(DownloadBatchState batch_state, int64_t query_version,
+                                                   bool same_remote_versions)
 {
     // Should never be called if session is not active
     REALM_ASSERT_EX(m_state == State::Active, m_state);
-    if (batch_state == DownloadBatchState::SteadyState) {
-        return true;
+    // Query version should always be the same or increasing
+    REALM_ASSERT_3(query_version, >=, m_wrapper.m_flx_active_version);
+
+    // Return early if already steady state or PBS (doesn't use bootstraps)
+    if (batch_state == DownloadBatchState::SteadyState || !m_is_flx_sync_session) {
+        return true; // Steady state
     }
 
-    if (!m_is_flx_sync_session) {
-        return true;
+    // Bootstrap messages (i.e. non-steady state) are identified by:
+    // * DownloadBatchState of MoreToCome
+    // * DownloadBatchState of LastInBatch, and
+    //   * first LastInBatch=true after one or more MoreToCome messages
+    //   * query_version greater than the active query_version
+    //   * LastInBatch=true message has 2 or more changesets and all have the same
+    //     remote_version
+
+    // LastInBatch=False messages are always a bootstrap message
+    if (batch_state == DownloadBatchState::MoreToCome) {
+        return false; // Not steady state
     }
 
-    // If this is a steady state DOWNLOAD, no need for special handling.
-    if (batch_state == DownloadBatchState::LastInBatch && query_version == m_wrapper.m_flx_active_version) {
-        return true;
+    // Messages with query_version greater than the active version are always a
+    // bootstrap message
+    if (query_version > m_wrapper.m_flx_active_version) {
+        return false; // Not steady state
+    }
+
+    // If this is the first LastInBatch=true message after one or more
+    // LastInBatch=false messages, this is the end of a bootstrap
+    if (m_last_download_batch_state == DownloadBatchState::MoreToCome) {
```

Comment:
i wish we didn't have to add this state tracking, but I think I see why we need it.